pkg/stub/handler_test.go

```diff
@@ -451,6 +453,89 @@ func TestImageStreamEvent(t *testing.T) {
 	validate(true, err, "", cfg, conditions, statuses, t)
 }
 
+func TestImageStreamErrorRetry(t *testing.T) {
+	h, cfg, event := setup()
+	processCred(&h, cfg, t)
+	mimic(&h, x86OCPContentRootDir)
+	err := h.Handle(event)
+	statuses := []corev1.ConditionStatus{corev1.ConditionFalse, corev1.ConditionTrue, corev1.ConditionTrue, corev1.ConditionTrue, corev1.ConditionFalse, corev1.ConditionFalse, corev1.ConditionFalse}
+	validate(true, err, "", cfg, conditions, statuses, t)
+	// expedite the stream events coming in
+	cache.ClearUpsertsCache()
+
+	tagVersion := int64(1)
+	is := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "foo",
+			Annotations: map[string]string{
+				v1.SamplesVersionAnnotation: h.version,
+			},
+		},
+		Spec: imagev1.ImageStreamSpec{
+			Tags: []imagev1.TagReference{
+				{
+					Name:       "foo",
+					Generation: &tagVersion,
+					From: &corev1.ObjectReference{
+						Kind: "DockerImage",
+						Name: "foofoo",
+					},
+				},
+			},
+		},
+		Status: imagev1.ImageStreamStatus{
+			Tags: []imagev1.NamedTagEventList{
+				{
+					Tag: "foo",
+					Conditions: []imagev1.TagEventCondition{
+						{
+							Generation: tagVersion,
+							Status: corev1.ConditionFalse,
+							Message: "failed import",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	h.processImageStreamWatchEvent(is, false)
+
+	if !cfg.ConditionTrue(v1.ImportImageErrorsExist) {
+		t.Fatalf("Import Error Condition not true: %#v", cfg)
+	}
+
+	importError := cfg.Condition(v1.ImportImageErrorsExist)
+	importErrorLastUpdateTime := cfg.Condition(v1.ImportImageErrorsExist).LastUpdateTime
+	h.processImageStreamWatchEvent(is, false)
+	if !importError.LastUpdateTime.Equal(&importErrorLastUpdateTime) {
+		t.Fatalf("Import Error Condition updated too soon: old update time %s new update time %s", importErrorLastUpdateTime.String(), importError.LastUpdateTime.String())
+	}
+
+	importError.LastUpdateTime.Time = metav1.Now().Add(-15 * time.Minute)
+	cfg.ConditionUpdate(importError)
+
+	h.processImageStreamWatchEvent(is, false)
+
+	if cfg.Condition(v1.ImportImageErrorsExist).LastUpdateTime.Equal(&importErrorLastUpdateTime) {
```

Comment:
conversely, this condition could never be false.  You set:

```
importError.LastUpdateTime.Time = metav1.Now().Add(-15 * time.Minute)
cfg.ConditionUpdate(importError)
```

So you've explicitly made the condition's lastupdate time different from the `importErrorLastUpdateTime` value you stored, so they'll never be equal, even if `processImageStreamWatchEvent(is, false)` doesn't update it.

Reply:
ok yep ... I've added some refetch / saves ... ptal